### PR TITLE
 Enhancements for circular genome viewer

### PR DESCRIPTION
**Pull request by maovt**
Please also include the updated the circulus submodule for new builds

### CODE REVIEW

File: public/js/p3/widget/CircularViewerContainer.js [hunk 1]
```diff
@@ -123,6 +124,95 @@ define([
...
 		},
 
+		addSpGeneTrack: function(title, title_tooltip, gid, filter, strand, fill, stroke, background){
+			var fields = ["feature_id", "feature_type", "sequence_id", "segments", "gi", "na_length", "pos_group", "strand", "public", "aa_length", "patric_id", "owner",
+				"location", "protein_id", "refseq_locus_tag", "taxon_id", "accession", "end", "genome_name", "product", "genome_id", "annotation", "start"];
+				
+			var sp_fields = ["feature_id","patric_id", "evidence", "property", "source"]
+
+			//var sp_query = "?eq(genome_id," + this.state.genome_ids[0] + ")&in(property,(%22Antibiotic%20Resistance%22,%22Virulence%20Factor%22,%22Transporter%22,%22Essential%20Gene%22))&sort(+patric_id,+property)" + "&select(" + sp_fields.join(",") + ")&limit(25000)";
+			var sp_query = "?eq(genome_id," + this.state.genome_ids[0] + ")" + filter + "&sort(+patric_id,+property)" + "&select(" + sp_fields.join(",") + ")&limit(25000)";
```
Comment (hyoo) [inline public/js/p3/widget/CircularViewerContainer.js]: this should work, but a little bit inefficient. 

1. you will not need anything more than feature_id in selection. that is what you're using in line 176
2. do you need to sort? i guess your filter already narrow down one property at a time.

File: public/js/p3/widget/CircularViewerContainer.js [hunk 2]
```diff
@@ -123,6 +124,95 @@ define([
...
+						feature_ids.push(spgenes[i].feature_id);
+					}
+					
+					var query = "in(feature_id,(" + feature_ids.join(",") + "))&sort(+accession,+start)" + "&select(" + fields.join(",") + ")&limit(25000)";
+					// console.log(" query:",  PathJoin(this.apiServiceUrl, "genome_feature", query));								 
+					
+					xhr.post(PathJoin(this.apiServiceUrl, "genome_feature/"), {
+						data: query,
+						headers: {
+							accept: "application/solr+json",
```
Comment (hyoo) [inline public/js/p3/widget/CircularViewerContainer.js]: if you request "application/json" as you did in line 161, you will get docs directly. then you don't need line 193

File: public/js/p3/widget/CircularViewerContainer.js [hunk 3]
```diff
@@ -123,6 +124,95 @@ define([
...
+						if (refseqs.length == 0) {
+							track.set('loading', false);
+							Topic.publish("/Notification", {message: "No data found.", type: "error"});                                          
+					
+							return refseqs;
+						}
+						// console.log(" refseqs:", refseqs); 										
+						refseqs = refseqs.map(function(r){
+							r.name = r.accession;
+							r.length = r.end - r.start;
```
Comment (hyoo) [inline public/js/p3/widget/CircularViewerContainer.js]: i think we do have na_length as a property of feature. you don't need to calculate on the fly

File: public/js/p3/widget/CircularViewerContainer.js [hunk 4]
```diff
@@ -161,8 +251,14 @@ define([
 			}, "outer", true);
 
 			//this.addFeatureTrack("CDS - FWD", "CDS forward strand", this.state.genome_ids[0], "and(eq(annotation,PATRIC),eq(feature_type,CDS))", null, "blue", null);
-			this.addFeatureTrack("CDS - FWD", "CDS forward strand", this.state.genome_ids[0], "and(eq(annotation,PATRIC),eq(feature_type,CDS),eq(strand,%22\+%22))", true, "#307D32", null)
-			this.addFeatureTrack("CDS - REV", "CDS reverse strand", this.state.genome_ids[0], "and(eq(annotation,PATRIC),eq(feature_type,CDS),eq(strand,%22-%22))", false, "#833B76", null)
+			this.addFeatureTrack("CDS - FWD", "CDS forward strand", this.state.genome_ids[0], "and(eq(annotation,PATRIC),eq(feature_type,CDS),eq(strand,%22\+%22))", true, "#307D32", null);
+			this.addFeatureTrack("CDS - REV", "CDS reverse strand", this.state.genome_ids[0], "and(eq(annotation,PATRIC),eq(feature_type,CDS),eq(strand,%22-%22))", false, "#833B76", null);
+			this.addSpGeneTrack("AMR Genes", "AMR Genes", this.state.genome_ids[0], "&in(property,(%22Antibiotic%20Resistance%22))", null, "red", null);
```
Comment (hyoo) [inline public/js/p3/widget/CircularViewerContainer.js]: isn't this equivalent to 
```eq(property,Antibiotic Resistance)```

File: public/js/p3/widget/CircularViewerContainer.js (hunk 1 above)
Comment (maovt) [reply public/js/p3/widget/CircularViewerContainer.js]: This was done before deciding only one property per track. The original plan was to have multiple proteries on one track.

Comment (hyoo) [reply public/js/p3/widget/CircularViewerContainer.js]: so, can we clean up then?

File: public/js/p3/widget/CircularViewerContainer.js (hunk 4 above)
Comment (maovt) [reply public/js/p3/widget/CircularViewerContainer.js]: Again, this was used for multiple properties before

File: public/js/p3/widget/CircularViewerContainer.js (hunk 2 above)
Comment (maovt) [reply public/js/p3/widget/CircularViewerContainer.js]: sure. I can change this

File: public/js/p3/widget/CircularViewerContainer.js (hunk 3 above)
Comment (maovt) [reply public/js/p3/widget/CircularViewerContainer.js]: Here we use nt length

File: public/js/p3/widget/CircularViewerContainer.js (hunk 4 above)
Comment (maovt) [reply public/js/p3/widget/CircularViewerContainer.js]: I can change this

File: public/js/p3/widget/CircularViewerContainer.js (hunk 1 above)
Comment (maovt) [reply public/js/p3/widget/CircularViewerContainer.js]: clean up and committed the changes

File: public/js/p3/widget/CircularViewerContainer.js (hunk 2 above)
Comment (maovt) [reply public/js/p3/widget/CircularViewerContainer.js]: Modified

File: public/js/p3/widget/CircularViewerContainer.js (hunk 3 above)
Comment (maovt) [reply public/js/p3/widget/CircularViewerContainer.js]: Kept as is.

**Pull request merged**